File: scripts/start-servers.pl
#!/usr/bin/perl

die "usage: ./start-servers.pl config_prefix" unless @ARGV==1;
my $configPrefix = shift;
my $configFile = "${configPrefix}0.config";
my $tssConfigFile = "${configPrefix}.tss.config";

system("../platform/build/server -c $configFile -i 0 &");
system("../platform/build/server -c $configFile -i 1 &");
system("../platform/build/server -c $configFile -i 2 &");

system("../platform/build/tss -c $tssConfigFile -i 0 &");
system("../platform/build/tss -c $tssConfigFile -i 1 &");
system("../platform/build/tss -c $tssConfigFile -i 2 &");

File: scripts/start-servers.py
#!/usr/bin/python

import argparse
import os
import re
from pyrem.host import RemoteHost

USER_NAME = "nl35"
BUILD_DIR = "/homes/sys/nl35/research/diamond-src/platform/build"

WORKING_DIR = "/biggerraid/users/" + USER_NAME + "/scratch"

parser = argparse.ArgumentParser(description='Launch servers.')
parser.add_argument('config_prefix', help='the config file prefix')
args = parser.parse_args()

serverConfigName = args.config_prefix + '0.config'
serverConfig = open(serverConfigName, 'r')

remoteServerConfigName = WORKING_DIR + "/diamond.0.config"

i = 0

for line in serverConfig:
    match = re.match("replica\s+([\w\.]+):(\d)", line)
    if match:
        hostname = match.group(1)

        os.system("scp " + serverConfigName + " " + USER_NAME + "@" + hostname + ":" + remoteServerConfigName)

        host = RemoteHost(hostname)
        command = BUILD_DIR + "/server -c " + remoteServerConfigName + " -i " + repr(i)
        server = host.run([command])
        server.start()

        i = i + 1
